e2e/LongHaul/service/DeviceOperations.cs

```diff
@@ -0,0 +1,138 @@
+                            (DateTimeOffset.UtcNow - responsePayload.CurrentTimeUtc).TotalSeconds);
+                    }
+
+                    _logger.Trace($"Response status: {response.Status}, payload:\n\t{JsonConvert.SerializeObject(response.PayloadAsString)}", TraceSeverity.Information);
+                }
+                catch (IotHubServiceException ex) when (ex.ErrorCode == IotHubServiceErrorCode.DeviceNotOnline)
+                {
+                    _logger.Trace($"Caught exception invoking direct method {ex}", TraceSeverity.Warning);
+                }
+
+                await Task.Delay(s_directMethodInvokeInterval, ct).ConfigureAwait(false);
+            }
+        }
+
+        public async Task SetDesiredPropertiesAsync(string keyName, string properties, CancellationToken ct)
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                var twin = new ClientTwin();
+                twin.Properties.Desired[keyName] = properties;
+
+                ++_totalDesiredPropertiesUpdatesCount;
+                _logger.Trace($"Updating the desired properties for device: {_deviceId}", TraceSeverity.Information);
+                _logger.Metric(TotalDesiredPropertiesUpdatesCount, _totalDesiredPropertiesUpdatesCount);
+
+                await _serviceClient.Twins.UpdateAsync(_deviceId, twin, false, ct).ConfigureAwait(false);
+
+                await Task.Delay(s_desiredPropertiesSetInterval, ct).ConfigureAwait(false);
+            }
+        }
+
+        public async Task SendC2dMessagesAsync(CancellationToken ct)
+        {
+            try
+            {
+                await _serviceClient.Messages.OpenAsync(ct).ConfigureAwait(false);
+
+                while (!ct.IsCancellationRequested)
+                {
+                    var payload = new CustomC2dMessagePayload
+                    {
+                        RandomId = Guid.NewGuid(),
+                        CurrentTimeUtc = DateTime.UtcNow,
+                        MessagesSentCount = ++_totalC2dMessagesSentCount,
+                    };
+                    var message = new OutgoingMessage(payload)
+                    {
+                        // An acknowledgment is sent on delivery success or failure.
+                        Ack = DeliveryAcknowledgement.Full,
+                        MessageId = payload.RandomId.ToString(),
+                    };
+
+                    _logger.Trace($"Sending message with Id {message.MessageId} to the device: {_deviceId}", TraceSeverity.Information);
+                    _logger.Metric(TotalC2dMessagesSentCount, _totalC2dMessagesSentCount);
+
+                    await _serviceClient.Messages.SendAsync(_deviceId, message, ct).ConfigureAwait(false);
+
+                    await Task.Delay(s_c2dMessagesSentInterval, ct).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                await _serviceClient.Messages.CloseAsync().ConfigureAwait(false);
```

Comment:
If this is a shared service client instance, you won't want to close this sub client.